### PR TITLE
Prevent nodes from gossiping with themselves with different ids

### LINKED ISSUE #689: Don't let validators gossip with themselves
STR:
```
$ ./multinode-demo/setup.sh -p 
$ ./multinode-demo/validator.sh
^C
$ ./multinode-demo/setup.sh -p 
$ ./multinode-demo/validator.sh
```

The second validator comes up on the same IP:port but with a different public key, and gets told about the dead first validator from other nodes.

Now the filter at https://github.com/solana-labs/solana/blob/master/src/crdt.rs#L740 doesn't exclude the first validator from gossip since the public key is different.   This causes a `RequestUpdates` to the first/dead validator [1] but that ultimately hits the second/live validator (since they had the same IP:port).   The second validator services the `RequestUpdates` and hits this code, https://github.com/solana-labs/solana/blob/master/src/crdt.rs#L433-L441, since the updates are identical to what it already has (cause it just sent it to itself!).  This prevents all nodes the validator is holding in its crdt from ever getting purged.

[1] https://github.com/solana-labs/solana/blob/master/src/crdt.rs#L764

### PR DESCRIPTION
**Pull request by mvines**
Fixes #689

### CODE REVIEW

**Comment by mvines** [comment]
(wip: needs a test...)

File: src/crdt.rs [hunk 1]
```diff
@@ -993,35 +993,48 @@ impl Crdt {
...
-                let len = data.len();
-                let rsp = Protocol::ReceiveUpdates(from, ups, data, external_liveness);
-                obj.write().unwrap().insert(&from_rd);
-                if len < 1 {
-                    let me = obj.read().unwrap();
-                    trace!(
-                        "no updates me {:x} ix {} since {}",
+                if addr == me.table[&me.me].contact_info.ncp {
+                    warn!(
+                        "RequestUpdates ignored, I'm talking to myself: me={:x} remoteme={:x}",
```
Comment (aeyakovenko) [inline src/crdt.rs]: can you add a counter for this

File: src/crdt.rs [hunk 2]
```diff
@@ -993,14 +993,27 @@ impl Crdt {
                 let addr = from_rd.contact_info.ncp;
                 trace!("RequestUpdates {} from {}", v, addr);
                 let me = obj.read().unwrap();
+                if addr == me.table[&me.me].contact_info.ncp {
+                    warn!(
+                        "RequestUpdates ignored, I'm talking to myself: me={:x} remoteme={:x}",
+                        me.debug_id(),
+                        make_debug_id(&from_rd.id)
+                    );
+                    inc_new_counter!("crdt-window-request-loopback", 1);
```
Comment (mvines) [inline src/crdt.rs]: Added counter

**Comment by mvines** [comment]
Tests added, r?

File: src/crdt.rs [hunk 3]
```diff
@@ -1370,12 +1394,25 @@ mod tests {
         assert_eq!(d.version, 0);
         let mut crdt = Crdt::new(d.clone()).unwrap();
         assert_eq!(crdt.table[&d.id].version, 0);
+        assert!(crdt.alive.get(&d.id).is_none());
```
Comment (garious) [inline src/crdt.rs]: `get(x).is_none()` -> `!contains()`  x 3

Comment (mvines) [reply src/crdt.rs]: ✔️  Looks like it's called [contains_key](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.contains_key)

**Comment by garious** [comment]
👍

**Comment by aeyakovenko** [comment]
Lgtm

**Comment by aeyakovenko** [comment]
@mvines Can you run the multinode dynamic test on 96core machine?

**Pull request merged**